### PR TITLE
[core] Start the synchronization connection after receiving all nodes info.

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <74173148+iycheng@users.noreply.github.com>

## Why are these changes needed?
There is a race condition between resource broadcasting and node registration. When the node is registered, it'll start the sync connection immediately. If the first message from the GCS is received before the Raylet got all node infos, it's going to drop the message from the node which it doesn't know.

The newly started raylet will only see its local resource which might cause some problem if it's going to schedule some task around.

This PR move the connection to the place after getting all nodes infos.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -441,10 +441,44 @@ ray::Status NodeManager::RegisterGcs() {
...
+      // Register a commands channel.
+      // It's only used for GC right now.
+      ray_syncer_.Register(
+          /* message_type */ syncer::MessageType::COMMANDS,
+          /* reporter */ this,
+          /* receiver */ this,
+          /* pull_from_reporter_interval_ms */ 0);
+
+      auto gcs_channel = gcs_client_->GetGcsRpcClient().GetChannel();
+      ray_syncer_.Connect(kGCSNodeID.Binary(), gcs_channel);
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:465]: Q: When we first connect to the GCS, does GCS update the snaphost resource usage of that node?

**Review by rkooo567 — APPROVED** [review]
Maybe let's try running microbenchmark + many_tasks to see the impact?

File: src/ray/raylet/node_manager.cc (hunk 1 above)
Comment (iycheng) [reply src/ray/raylet/node_manager.cc:465]: I think this logic is the same as before. It's not pushed to GCS immediately but will wait until first message generated (100ms). First pull will always succeeds!

But you got a good point here. I double checked the code. When node register it'll push its local resource to GCS. And when node added, it'll use that to update its local view.


I think my unit test might not work here. Let me double check.

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -441,10 +441,44 @@ ray::Status NodeManager::RegisterGcs() {
...
+            }
+          },
+          RayConfig::instance().raylet_check_gc_period_milliseconds(),
+          "NodeManager.CheckGC");
+    }
+  };
   // Register a callback to monitor new nodes and a callback to monitor removed nodes.
-  RAY_RETURN_NOT_OK(
-      gcs_client_->Nodes().AsyncSubscribeToNodeChange(on_node_change, on_done));
+  RAY_RETURN_NOT_OK(gcs_client_->Nodes().AsyncSubscribeToNodeChange(
```
Comment (jjyao) [inline src/ray/raylet/node_manager.cc:480]: For future nodes, is it guaranteed that node notification arrives before node resource update?

**Review by jjyao — APPROVED** [review]

File: src/ray/raylet/node_manager.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/raylet/node_manager.cc:480]: That's a good point. I'll comeback to this PR later.

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -1191,11 +1201,6 @@ void NodeManager::HandleNotifyGCSRestart(rpc::NotifyGCSRestartRequest request,
 
 bool NodeManager::UpdateResourceUsage(const NodeID &node_id,
                                       const rpc::ResourcesData &resource_data) {
-  // Trigger local GC at the next heartbeat interval.
```
Comment (iycheng) [inline src/ray/raylet/node_manager.cc:1194]: Not sure why it's added here. It seems being added by gcs_actor_scheduler which is not enabled by default. I don't think it's the right place to add this. Delete and check CI tests.

**Pull request merged**